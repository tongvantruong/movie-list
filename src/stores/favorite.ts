import { LOCAL_KEY_FAVORITES } from '@/const/key'
import type { Movie } from '@/models/Movie'
import {
  createGlobalState,
  StorageSerializers,
  useLocalStorage,
  type RemovableRef,
} from '@vueuse/core'
import { computed, type ComputedRef } from 'vue'

type FavoriteMovies = Movie[]
const defaultMovies: FavoriteMovies = []

export const favoriteStore = createGlobalState(() => {
  const storage: RemovableRef<FavoriteMovies> = useLocalStorage(
    LOCAL_KEY_FAVORITES,
    defaultMovies,
    {
      serializer: StorageSerializers.object,
    },
  )

  const favoriteMovies: ComputedRef<FavoriteMovies> = computed(() => storage.value)

  function like(movie: Movie) {
    storage.value.push(movie)
  }

  function unlike(imdbId: Movie['imdbId']) {
    storage.value = storage.value.filter((it) => it.imdbId !== imdbId)
  }

  function isLiked(imdbId: Movie['imdbId']): boolean {
    return !!storage.value.find((it) => it.imdbId === imdbId)
  }

  return { like, unlike, isLiked, favoriteMovies }
})
