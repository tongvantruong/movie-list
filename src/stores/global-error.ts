import { createGlobalState } from '@vueuse/core'
import { ref, type Ref } from 'vue'

export const globalErrorStore = createGlobalState(() => {
  const show: Ref<boolean> = ref(false)

  const showError = () => {
    show.value = true
  }

  return { show, showError }
})
