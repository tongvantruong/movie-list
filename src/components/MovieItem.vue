<template>
  <VCard
    class="movie-item card-shadow"
    role="listitem"
    @dblclick="toggleStar"
    v-ripple="{ class: 'text-warn' }"
  >
    <template #title>
      <div class="d-flex justify-space-between ga-2">
        <span class="movie-item__title">{{ movie.title }}</span>
        <VTooltip location="bottom" role="tooltip" aria-label="favorite">
          <template v-slot:activator="{ props }">
            <VBtn
              title="Star"
              class="movie-item__icon"
              data-cy="movie-item-icon"
              v-bind="props"
              :icon="iconStar"
              variant="text"
              @click="toggleStar"
            />
          </template>
          <span data-cy="movie-item-tooltip" v-html="tooltipForStarIcon"></span>
        </VTooltip>
      </div>
    </template>
    <template #subtitle>
      <div class="d-flex justify-space-between ga-2">
        <span test-id="movie-item-year">
          Year: <i>{{ movie.year }}</i>
        </span>
        <span test-id="movie-item-imdbid">
          IMDb ID: <i>{{ movie.imdbId }}</i>
        </span>
      </div>
    </template>
  </VCard>
</template>

<script setup lang="ts">
import type { Movie } from '@/models/Movie'
import { favoriteStore } from '@/stores/favorite'
import { computed, ref, type PropType, type ComputedRef, type Ref } from 'vue'
import { useI18n } from 'vue-i18n'

const { t } = useI18n()

const { movie } = defineProps({
  movie: { type: Object as PropType<Movie>, required: true },
})

const { like, unlike, isLiked } = favoriteStore()

const isStarred: Ref<boolean> = ref(isLiked(movie.imdbId))

const iconStar: ComputedRef<string> = computed(() =>
  isStarred.value ? 'mdi-star' : 'mdi-star-outline',
)

const tooltipForStarIcon: ComputedRef<string> = computed(() => {
  return t('favorite_tooltip')
})

function toggleStar() {
  isStarred.value = !isStarred.value

  if (isStarred.value) like(movie)
  else unlike(movie.imdbId)
}
</script>

<style lang="scss" scoped>
.movie-item {
  width: 100%;

  &:hover {
    background-color: $colorHover;
  }

  &__icon {
    color: $colorPrimary;
  }

  &__title {
    display: -webkit-box;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
    overflow: hidden;
    text-overflow: ellipsis;
    text-wrap: wrap;
    text-align: start;
  }
}
</style>
