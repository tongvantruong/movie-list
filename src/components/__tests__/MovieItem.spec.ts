import { describe, it, expect, vi, afterEach } from 'vitest'
import { createVuetify } from 'vuetify'
import { mount, VueWrapper } from '@vue/test-utils'
import MovieItem from '@/components/MovieItem.vue'
import { beforeEach } from 'vitest'
import type { Movie } from '@/models/Movie'
import * as favoriteStore from '@/stores/favorite'

const useFavoritesMock = vi.hoisted(() => ({
  like: vi.fn(),
  unlike: vi.fn(),
  isLiked: vi.fn(),
}))

const vuetify = createVuetify()

let wrapper: VueWrapper

const movie: Movie = {
  title: 'One Piece',
  year: 1999,
  imdbId: 'tt0388629',
}

function mountComponent() {
  wrapper = mount(MovieItem, {
    props: { movie },
    global: {
      plugins: [vuetify],
    },
  })
}

beforeEach(() => {
  mountComponent()
})

afterEach(() => {
  vi.restoreAllMocks()
})

describe('MovieItem', () => {
  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should render title', () => {
    const title = wrapper.find('.movie-item__title')
    expect(title.text()).toBe(movie.title)
  })
  it('should render year', () => {
    const year = wrapper.find('[test-id="movie-item-year"]')
    expect(year.text()).toBe(`Year: ${movie.year}`)
  })
  it('should render Imdb ID', () => {
    const imdbId = wrapper.find('[test-id="movie-item-imdbid"]')
    expect(imdbId.text()).toBe(`IMDb ID: ${movie.imdbId}`)
  })
  it('should change stared ref if clicked star icon', async () => {
    const icon = wrapper.find('.movie-item__icon')
    // @ts-expect-error to access stared ref
    expect(wrapper.vm.isStarred).toBe(false)
    await icon.trigger('click')
    // @ts-expect-error to access stared ref
    expect(wrapper.vm.isStarred).toBe(true)

    await icon.trigger('click')
    // @ts-expect-error to access stared ref
    expect(wrapper.vm.isStarred).toBe(false)
  })
  it('should call "star" with correct movie', async () => {
    // @ts-expect-error to test
    vi.spyOn(favoriteStore, 'favoriteStore').mockReturnValue(useFavoritesMock)
    mountComponent()

    const icon = wrapper.find('.movie-item__icon')
    await icon.trigger('click')
    expect(useFavoritesMock.like).toHaveBeenCalledTimes(1)
    expect(useFavoritesMock.like).toHaveBeenCalledWith(movie)
  })

  it('should call "unStar" with correct movie', async () => {
    // @ts-expect-error to test
    vi.spyOn(favoriteStore, 'favoriteStore').mockReturnValue(useFavoritesMock)
    mountComponent()

    const icon = wrapper.find('.movie-item__icon')
    await icon.trigger('click')
    await icon.trigger('click')
    expect(useFavoritesMock.unlike).toHaveBeenCalledTimes(1)
    expect(useFavoritesMock.unlike).toHaveBeenCalledWith(movie.imdbId)
  })
})
