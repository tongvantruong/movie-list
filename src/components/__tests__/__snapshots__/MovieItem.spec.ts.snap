// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MovieItem > should match snapshot 1`] = `
"<div data-v-a5f46965="" class="v-card v-theme--light v-card--density-default v-card--variant-elevated movie-item card-shadow" role="listitem">
  <!---->
  <div class="v-card__loader">
    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
      <!---->
      <div class="v-progress-linear__background"></div>
      <div class="v-progress-linear__buffer" style="width: 0%;"></div>
      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <div class="v-card-item">
    <!---->
    <div class="v-card-item__content">
      <div class="v-card-title">
        <div data-v-a5f46965="" class="d-flex justify-space-between ga-2"><span data-v-a5f46965="" class="movie-item__title">One Piece</span><button data-v-a5f46965="" type="button" class="v-btn v-btn--icon v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-text movie-item__icon" title="Star" data-cy="movie-item-icon" aria-describedby="v-tooltip-0"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
            <!----><span class="v-btn__content" data-no-activator=""><i class="mdi-star-outline mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span>
            <!---->
            <!---->
          </button>
          <!--teleport start-->
          <!--teleport end-->
        </div>
      </div>
      <div class="v-card-subtitle">
        <div data-v-a5f46965="" class="d-flex justify-space-between ga-2"><span data-v-a5f46965="" test-id="movie-item-year"> Year: <i data-v-a5f46965="">1999</i></span><span data-v-a5f46965="" test-id="movie-item-imdbid"> IMDb ID: <i data-v-a5f46965="">tt0388629</i></span></div>
      </div>
      <!---->
    </div>
    <!---->
  </div>
  <!---->
  <!---->
  <!---->
  <!----><span class="v-card__underlay"></span>
</div>"
`;
